File: edtc_matlab/srs.m
% Filename: srs.m
% Author: Andy Qi
% Date: December 2008
% Corresponds to: Listing 6.1

classdef srs < handle
    properties
        % Parameters : F and phi are functions, where
        % phi() returns a draw from phi. X is a
        % number representing the initial condition
        F;
        phi;
        X;
    end
    
    methods
        function self = srs(F,phi,X)
            % Represents X_{t + 1} = F(X_t,W_{t + 1}); W ~ phi
            self.F = F;
            self.phi = phi;
            self.X = X;
        end
        
        function update(self)
            % Update the state according to X = F(X, W)
            self.X = self.F(self.X,self.phi());
        end
        
        function path = sample_path(self,n)
            % Generate path of length n from current state
            path = zeros(1,n);
            for i = 1:n
                path(i) = self.X;
                self.update;
            end
        end
    end
end

File: edtc_python/srs.py
# Filename: srs.py
# Author: John Stachurski
# Date: December 2008
# Corresponds to: Listing 6.1

class SRS:
    
    def __init__(self, F=None, phi=None, X=None):
        """Represents X_{t+1} = F(X_t, W_{t+1}); W ~ phi.
        Parameters: F and phi are functions, where phi() 
        returns a draw from phi. X is a number representing 
        the initial condition."""
        self.F, self.phi, self.X = F, phi, X

    def update(self):
        "Update the state according to X = F(X, W)."
        self.X = self.F(self.X, self.phi())

    def sample_path(self, n):
        "Generate path of length n from current state."
        path = []
        for i in range(n):
            path.append(self.X)
            self.update()
        return path
